test/infrastructure/docker/exp/internal/docker/nodepool.go

```diff
@@ -168,15 +169,23 @@ func (np *NodePool) Delete(ctx context.Context) error {
 }
 
 func (np *NodePool) isMachineMatchingInfrastructureSpec(machine *docker.Machine) bool {
-	return imageVersion(machine) == container.SemverToOCIImageTag(*np.machinePool.Spec.Template.Spec.Version)
-}
+	// NOTE: with the current implementation we are checking if the machine is using a kindest/node image for the expected version,
+	// but not checking if the machine has the expected extra.mounts or pre.loaded images.
+
+	// If the machine has been created with a custom image, check if we are using it.
+	if np.dockerMachinePool.Spec.Template.CustomImage != "" {
+		return machine.ContainerImage() == np.dockerMachinePool.Spec.Template.CustomImage
+	}
+
+	// Otherwise check if the machine is using the kind image we resolve to for a given Kubernetes version:
+	semVer, err := semver.Parse(strings.TrimPrefix(*np.machinePool.Spec.Template.Spec.Version, "v"))
+	if err != nil {
+		panic(errors.Wrap(err, "failed to parse DockerMachine version").Error())
```

Comment:
Can we add a todo to return error instead of panic'ing in a follow-up? I assume this is a "should never happen" case, but just panic'ing is still not great